src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -299,12 +346,16 @@ public virtual bool TryHandleTab()
         {
             if (ExpansionSession != null)
             {
-                var tabbedInsideSnippetField = VSConstants.S_OK == ExpansionSession.GoToNextExpansionField(0);
+                // When 'Tab' is pressed in the last field of a normal snippet, the session wraps back around to the
+                // first field (this is preservation of historical behavior). When 'Tab' is pressed at the end of an
+                // argument provider snippet, the snippet session is automatically committed (this behavior matches the
+                // design for Insert Full Method Call intended for multiple IDEs).
+                var tabbedInsideSnippetField = VSConstants.S_OK == ExpansionSession.GoToNextExpansionField(fCommitIfLast: _state.IsFullMethodCallSnippet ? 1 : 0);
 
                 if (!tabbedInsideSnippetField)
                 {
                     ExpansionSession.EndCurrentExpansion(fLeaveCaret: 1);
-                    ExpansionSession = null;
+                    _state.Clear(forceClearSymbolInformation: true);
```

Comment:
i didn't realize State was mutable until this line.  My preference would either be:

1. state is mutable.  But then the field is readonly so that it is only updated by these mutation methods.
2. state is immutable.  Changing the state is done by overwriting the _state field.

Reply:
Totally agree

Reply:
> state is mutable. But then the field is readonly so that it is only updated by these mutation methods.

This is the current behavior. `State` is non-copyable so all operations occur on the single mutable instance.

Reply:
⏱ am converting this from a non-copyable struct to a class, though it has no real difference in behavior.

Reply:
It's not great, but leaving significant rewrite to a separate follow-up.